akka-projection-core/src/main/scala/akka/projection/scaladsl/StatefulHandler.scala

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.scaladsl
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
+
+import akka.Done
+import akka.annotation.ApiMayChange
+
+@ApiMayChange abstract class StatefulHandler[State, Envelope](implicit ec: ExecutionContext) extends Handler[Envelope] {
+
+  private var state: Future[State] = _
+
+  /**
+   * Invoked to load the initial state when the projection is started or if previous `process` failed.
+   */
+  def initialState(): Future[State]
+
+  /**
+   * The `process` method is invoked for each `Envelope`.
+   * One envelope is processed at a time. The returned `Future[State]` is to be completed when the processing
+   * of the `envelope` has finished. It will not be invoked with the next envelope until after the returned
+   * `Future` has been completed.
+   *
+   * The `state` is the completed value of the previously returned `Future[State]` or the `initialState`.
+   * If the previously returned `Future[State]` failed it will call `initialState`
+   * again and use that value.
+   */
+  def process(state: State, envelope: Envelope): Future[State]
+
+  /**
+   * Calls [[StatefulHandler.initialState]] when the projection is started.
+   */
+  final override def start(): Future[Done] = {
+    state = initialState()
+    state.map(_ => Done)
+  }
+
+  /**
+   * Calls `process(state, envelope)` with the completed value of the previously returned `Future[State]`
+   * or the `initialState`. If the previously returned `Future[State]` failed it will call `initialState`
+   * again and use that value.
```

Comment:
It would be nice if we could keep a copy of the last processed state that can be restored on failure (perhaps optionally if a retry strategy is being used), but I guess this would difficult to enforce since the user could just keep passing a reference to the same mutable state object every iteration.  We could _try_ to keep a reference to the last state and document what to do if they want to be able to rollback to a previous state.

Reply:
I don't think we should make this too advanced, there will always be more things missing and then we have ended up implementing an actor.

In one of the examples I use Future.recover for a failure scenario